### PR TITLE
Add OffsetParameter

### PR DESCRIPTION
**Pull request by jetuk**
I also fixed a bug that meant the `NegativeMinParameter` was not being tested and the `NegativeMaxParameter` was not being tested properly.

### CODE REVIEW

File: pywr/parameters/_parameters.pyx [hunk 1]
```diff
@@ -1577,6 +1577,58 @@ cdef class NegativeMinParameter(MinParameter):
 NegativeMinParameter.register()
 
 
+cdef class OffsetParameter(Parameter):
+    """Parameter that takes offsets another `Parameter` by a constant value.
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:1581]: Delete 'takes'

File: pywr/parameters/_parameters.pyx [hunk 2]
```diff
@@ -1577,6 +1577,58 @@ cdef class NegativeMinParameter(MinParameter):
...
+
+    This class is a more efficient version of `AggregatedParameter` where
+    a single `Parameter` is offset by a constant value.
+
+    Parameters
+    ----------
+    parameter : `Parameter`
+        The parameter to to compare with the float.
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:1589]: delete extra 'to'

File: pywr/parameters/_parameters.pyx [hunk 3]
```diff
@@ -1577,6 +1577,58 @@ cdef class NegativeMinParameter(MinParameter):
...
+    offset : float (default=0.0)
+        The offset to add to apply to value returned by `parameter`.
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:1591]: delete 'to add'

File: pywr/parameters/_parameters.pyx [hunk 4]
```diff
@@ -1577,6 +1577,58 @@ cdef class NegativeMinParameter(MinParameter):
...
+        The upper bounds of the offset when used during optimisation.
+    """
+    def __init__(self, model, parameter, offset=0.0, lower_bounds=0.0, upper_bounds=np.inf, *args, **kwargs):
+        super(OffsetParameter, self).__init__(model, *args, **kwargs)
+        self.parameter = parameter
+        self.children.add(parameter)
+        self.offset = offset
+        self.double_size = 1
+        self._lower_bounds = np.ones(self.double_size) * lower_bounds
+        self._upper_bounds = np.ones(self.double_size) * upper_bounds
```
Comment (Batch21) [inline pywr/parameters/_parameters.pyx:1604]: There are a few other parameters that need need to have these optimisation attributes setup in the same way (e.g. `MaxParameter`).

Could the parameters that have another parameter and a single float as input args all inherit from a single base class that defines these attributes and associated methods?

Comment (jetuk) [reply pywr/parameters/_parameters.pyx:1604]: I'm not sure how you would do this cleanly (e.g. this offset uses a different internal name). I think there could be a common base class for all of `MaxParameter`, `MinParameter`, `NegativeMaxParameter` and `NegativeMinParameter` to add this support. Probably one for another issue. It is also related to #632. What do you think?

Comment (jetuk) [reply pywr/parameters/_parameters.pyx:1604]: We should probably have a `MinMaxParameter` too.

**Comment by jetuk** [comment]
@Batch21 I've fixed the docstring typos (please check) and replied to the other comment. Thanks.

File: pywr/parameters/_parameters.pyx (hunk 4 above)
Comment (Batch21) [reply pywr/parameters/_parameters.pyx:1604]: I agree it is a separate issue related to #632. Looks like there are quite a few parameters that people potentially might want to use as optimisation variables but are not currently setup with the new attributes.

**Pull request merged**